### PR TITLE
feat: allows to specify context to fetch git data

### PR DESCRIPTION
**Pull request by neilime**
Fix https://github.com/docker/metadata-action/issues/139

Add `context` input to specify where to get context data. Allowed options are: "workflow"  (default), "git".

#### `workflow` context

```yaml
context: workflow
```

```
sha: context.sha
ref: context.ref
```

#### `git` context

Will get Git context from previously checkout repository. @see https://github.com/actions/checkout

```yaml
context: git
```

```
sha: git rev rev-parse HEAD
ref: git symbolic-ref HEAD
```

### CODE REVIEW

**Comment by fredleger** [comment]
We also ran into an issue that can be fixed by this PR. From my point of view it makes totally sense

**Comment by crazy-max** [comment]
Thanks for your contrib, I think you can squash your commits and also fix DCO, see https://github.com/docker/metadata-action/pull/248/checks?check_run_id=10165522880

**Comment by neilime** [comment]
> Thanks for your contrib, I think you can squash your commits and also fix DCO, see https://github.com/docker/metadata-action/pull/248/checks?check_run_id=10165522880

Done @crazy-max

**Review by crazy-max — CHANGES_REQUESTED** [review]

**Comment by neilime** [comment]
@crazy-max needs your approval to run workflow and your review.

Thanks

**Comment by qwqcode** [comment]
This PR solves the problem I'm having perfectly, I would really appreciate it if you could merge it soon. 😁

File: package.json [hunk 1]
```diff
@@ -30,8 +30,9 @@
   "license": "Apache-2.0",
   "dependencies": {
     "@actions/core": "^1.10.0",
+    "@actions/exec": "^1.1.1",
```
Comment (crazy-max) [inline package.json:33]: not needed anymore

File: action.yml [hunk 2]
```diff
@@ -32,6 +32,10 @@ inputs:
     description: 'GitHub Token as provided by secrets'
     default: ${{ github.token }}
     required: true
+  context:
+    description: 'Where to get context data. Allowed options are "workflow"  (default), "git".'
+    default: "workflow"
+    required: true
```
Comment (crazy-max) [inline action.yml:38]: move on top (before `images`)

File: README.md [hunk 3]
```diff
@@ -320,6 +322,38 @@ So it can be used with our [Docker Build Push action](https://github.com/docker/
...
+
+```yaml
+context: git
+```
+
+```
+sha: git rev rev-parse HEAD
+ref: git symbolic-ref HEAD
+```
+
```
Comment (crazy-max) [inline README.md:356]: I don't think we need that level of detail. Would also prefer the same layout as other inputs. Something like this should enough:

<pre>
## `context` input

`context` defines where to get context metadata:

```yaml
# default
context: workflow
# or
context: git
```

* `workflow`: Get context metadata from the workflow (GitHub context). See https://docs.github.com/en/actions/learn-github-actions/contexts#github-context
* `git`: Get context metadata from the workflow and overrides some of them with current Git context, such as `ref` and `sha`.
</pre>

**Review by crazy-max — CHANGES_REQUESTED** [review]
Thanks for updating with latest actions-toolkit release. See my comments.

See also ci: https://github.com/docker/metadata-action/actions/runs/4323012276/jobs/7547788506#step:4:5637

**Comment by neilime** [comment]
@crazy-max, feedbacks are resolved and tests fixed, can you have a look please.

File: __tests__/meta.test.ts [hunk 4]
```diff
@@ -49,7 +58,8 @@ const tagsLabelsTest = async (name: string, envFile: string, inputs: Inputs, exV
   process.env = dotenv.parse(fs.readFileSync(path.join(__dirname, 'fixtures', envFile)));
   const toolkit = new Toolkit();
   const repo = await toolkit.github.repoData();
-  const meta = new Meta({...getInputs(), ...inputs}, new Context(), repo);
+  const context = await getFreshWorkflowContext();
+  const meta = new Meta({...getInputs(), ...inputs}, context, repo);
```
Comment (crazy-max) [inline __tests__/meta.test.ts:62]: I don't think and extra func is necessary:

```suggestion
  const meta = new Meta({...getInputs(), ...inputs}, await getContext(ContextSource.workflow), repo);
```

File: src/context.ts [hunk 5]
```diff
@@ -10,6 +13,7 @@ export interface Inputs {
   sepLabels: string;
   bakeTarget: string;
   githubToken: string;
+  context: ContextSource;
```
Comment (crazy-max) [inline src/context.ts:16]: Move on top before images

File: src/context.ts [hunk 6]
```diff
@@ -21,6 +25,49 @@ export function getInputs(): Inputs {
     sepTags: core.getInput('sep-tags', {trimWhitespace: false}) || `\n`,
     sepLabels: core.getInput('sep-labels', {trimWhitespace: false}) || `\n`,
     bakeTarget: core.getInput('bake-target') || `docker-metadata-action`,
-    githubToken: core.getInput('github-token')
+    githubToken: core.getInput('github-token'),
+    context: (core.getInput('context') || ContextSource.workflow) as ContextSource
```
Comment (crazy-max) [inline src/context.ts:29]: ditto

File: README.md [hunk 7]
```diff
@@ -22,6 +22,7 @@ ___
   * [inputs](#inputs)
   * [outputs](#outputs)
   * [environment variables](#environment-variables)
+* [`context` input](#tags-context)
```
Comment (crazy-max) [inline README.md:25]:
```suggestion
* [`context` input](#context-input)
```

**Review by crazy-max — CHANGES_REQUESTED** [review]
Some nits and should be good after that

File: __tests__/meta.test.ts (hunk 4 above)
Comment (neilime) [reply __tests__/meta.test.ts:62]: We are forced to use `getFreshWorkflowContext` as `Github.context` is loaded on import, so it does not use the `process.env` mocked variables

Comment (crazy-max) [reply __tests__/meta.test.ts:62]: Ah gotcha you need this to reset mocks

**Review by crazy-max — APPROVED** [review]
LGTM thanks!

**Pull request merged**